deprecation_notifier/DeprecationNotifier/DNAppDelegate.m

```diff
@@ -53,15 +64,38 @@ - (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
       expectedVersionArray = [expectedVersionArray arrayByAddingObject:@"0"];
   }
   
-    
+  NSLog(@"Info: System version: %@ %@", systemVersion, systemBuild);
+  NSLog(@"Info: Expected version: %@ %@", expectedVersion, expectedBuildsStr);
   if (systemVersionArray.count < 3 || expectedVersionArray.count < 3) {
     NSLog(@"Exiting: Error, unable to properly determine system version or expected version");
     [NSApp terminate:nil];
+  } else if ([expectedVersionArray[0] intValue] < [systemVersionArray[0] intValue]) {
+    NSLog(@"Exiting: OS is already %@ or greater", expectedVersion);
+    [NSApp terminate:nil];
+  } else if (([expectedVersionArray[0] intValue] <= [systemVersionArray[0] intValue]) &&
+             ([expectedVersionArray[1] intValue] <  [systemVersionArray[1] intValue])) {
+    NSLog(@"Exiting: OS is already %@ or greater", expectedVersion);
+    [NSApp terminate:nil];
   } else if (([expectedVersionArray[0] intValue] <= [systemVersionArray[0] intValue]) &&
              ([expectedVersionArray[1] intValue] <= [systemVersionArray[1] intValue]) &&
-             ([expectedVersionArray[2] intValue] <= [systemVersionArray[2] intValue])) {
+             ([expectedVersionArray[2] intValue] <  [systemVersionArray[2] intValue])) {
     NSLog(@"Exiting: OS is already %@ or greater", expectedVersion);
     [NSApp terminate:nil];
+  } else if (([expectedVersionArray[0] intValue] == [systemVersionArray[0] intValue]) &&
+             ([expectedVersionArray[1] intValue] == [systemVersionArray[1] intValue]) &&
+             ([expectedVersionArray[2] intValue] == [systemVersionArray[2] intValue])) {
+    NSLog(@"Checking: OS is equal to %@, checking build version", expectedVersion);
+    if (expectedBuilds == nil) {
+      NSLog(@"Exiting: No preferred build, so considered equal to %@", expectedVersion);
+      [NSApp terminate:nil];
+    } else {
+      if ([expectedBuilds containsObject:systemBuild]) {
+        NSLog(@"Exiting: OS build is one of the expected builds [ %@ ]", expectedBuildsStr);
+        [NSApp terminate:nil];
+      } else {
+        NSLog(@"Checking: OS build not found in expected builds [ %@ ]", expectedBuildsStr);
+      }
+    }
```

Comment:
This logic, while correct, is a bit more difficult to read as it's quite repetitive. Could we flip it around?

```objc
if ([expectedVersionArray isEqualToArray:systemVersionArray]) {
  NSLog(@"Checking: running OS is equal to %@, checking build version", expectedVersion);
  if (!expectedBuilds.count) {
    NSLog(@"Exiting: No preferred build, so considered equal");
    [NSApp terminate:nil];
  }
  if ([expectedBuilds containsObject:systemBuild]) {
    NSLog(@"Exiting: OS build is one of the expected builds %@", expectedBuilds);
    [NSApp terminate:nil];
  }
  NSLog(@"Checking: OS build not found in expected builds %@", expectedBuilds);
} else if ([systemVersionArray[0] intValue] < [expectedVersionArray[0] intValue] ||
           [systemVersionArray[1] intValue] < [expectedVersionArray[1] intValue] ||
           [systemVersionArray[2] intValue] < [expectedVersionArray[2] intValue]) {
  NSLog(@"Checking: OS build is lower than required");
} else {
  NSLog(@"Exiting: OS build is greater than required");
  [NSapp terminate:nil];
}
```